### DISCUSSION TITLE
Help needed to verify a custom toolkit

### DISCUSSION THREAD

**Discussion by yadavan88** [VirtusLab/scala-cli#2637] [General]
Hi Scala-CLI team,
I am trying to create a new toolkit. I followed the toolkit's github structure and created sample one. I also pushed the code to github [here](https://github.com/yadavan88/lihaoyi-toolkit).

After I created and published the toolkit locally using the command:
```
scala-cli --power publish local --cross LiHaoyiToolkit.scala publish-conf.scala
``` 
Now, I tried using the newly created toolkit as below:
```
//> using scala 3
//> using toolkit com.yadavan88:toolkit:0.1.0

import fansi.*
import upickle.default.macroRW
import upickle.default.*

case class Country(name: String, code: String, capital:String, languages: List[String])
object Country {
  implicit val rw: ReadWriter[Country] = macroRW
}

object Test {

  @main
  def main() = {
    val url = "https://raw.githubusercontent.com/yadavan88/blog-code-samples/main/countries.json"
    val countries = requests.get(url)
    println("Countries = "+countries.text())
    val countriesList = read[List[Country]](countries.text())
    println("No of countries = "+fansi.Color.Green(countriesList.size.toString))    
  }

}
```
I am able to execute this and run it successfully. However I have a few questions related to this. Could someone clarify this?

1. Is this attempt correct? 
2. Do I have to publish it as "toolkit" itself? I wanted to publish it as `lihaoyi-toolkit` instead, but it was not able to find the dependencies when I do that (change name in `publish.name`). 

Thanks in advance
Yadu

**Comment by tgodzik** [comment]
> Is this attempt correct?

Looks correct, though it's basically the same as adding a dependency and bringing other libraries as transitive dependencies.

> Do I have to publish it as "toolkit" itself? I wanted to publish it as lihaoyi-toolkit instead, but it was not able to find the dependencies when I do that (change name in publish.name).

@Gedochao would have to correct me, but I think that's how it got implemented in https://github.com/VirtusLab/scala-cli/pull/2025/files

I don't see the exact name being used, so probably it's hardcoded to `toolkit`

**Reply by yadavan88** [reply]
@tgodzik Thanks for the response. So, for the typelevel toolkit to be used as `using toolkit typelevel:latest`, a change was needed in the scala-cli core code, right? I thought it was somehow configurable somewhere. I was initially trying to understand how the typelevel toolkit was resolving(and thought that was the difference between publishing as a toolkit and using a library in terms of the syntax). 
So apart from the grouping of libs(and managing versions), what else can be done in a custom toolkit. I was just curious to know and try out, so created a custom toolkit. 
Thanks again!

**Reply by Gedochao** [reply]
> So, for the typelevel toolkit to be used as using toolkit typelevel:latest, a change was needed in the scala-cli core code, right? I thought it was somehow configurable somewhere.

Yup, that's a hardcoded alias right there. it's not configurable anywhere.
As noted in another comment, custom toolkits need to use the `org` there.

**Comment by Gedochao** [comment]
@yadavan88 
```scala
//> using toolkit com.yadavan88:toolkit:0.1.0
```
You can actually shorten it to just
```scala
//> using toolkit com.yadavan88:0.1.0
```

> 1. Is this attempt correct?

Yup, seems correct. I tested it locally and works as expected.
I suppose we are missing a section on creating custom toolkits in [the relevant guide](https://scala-cli.virtuslab.org/docs/guides/introduction/toolkit#other-toolkits), I created a ticket: https://github.com/VirtusLab/scala-cli/issues/2639
@yadavan88 feel free to fill it in based on your attempt if you find the time.

> 2. Do I have to publish it as "toolkit" itself? I wanted to publish it as lihaoyi-toolkit instead, but it was not able to find the dependencies when I do that (change name in publish.name).

Yes. The main scope artifact has to be named `toolkit`, while the test scope one has to be named `toolkit-test` (which is how it's [already configured in the linked repo](https://github.com/yadavan88/lihaoyi-toolkit/blob/main/LiHaoyiToolkitTest.scala#L2)).
So generally speaking, toolkits can be differentiated by org.
We also allow hardcoded aliases for popular ones (i.e. for [typelevel-toolkit](https://typelevel.org/toolkit/) it's sufficient to just pass `//> using toolkit typelevel:0.1.20` instead of `//> using toolkit org.typelevel:0.1.20`), but we do not support configuring similar aliases for custom toolkits at this time.

**Reply by yadavan88** [reply]
@Gedochao Thanks for the detailed clarification. I will try to fill up the details in the ticket. 
Could you also help me to understand some more cases where a toolkit makes sense than adding the libraries directly(apart from the obvious easier to manage versions)? Is there any other things that we can add in the toolkit that does some useful magic :D 
Just trying to get more clarity on it.

**Reply by yadavan88** [reply]
Also, I am planning on an intro talk on Scala CLI and toolkit in one of the meetups and trying to find out max info that I can share to make the talk more useful.

**Reply by Gedochao** [reply]
> Could you also help me to understand some more cases where a toolkit makes sense than adding the libraries directly(apart from the obvious easier to manage versions)? Is there any other things that we can add in the toolkit that does some useful magic :D

I suppose there's a little bit of magic, but then, not a lot.
Off the top of my head:
- as you mentioned, someone does the version management for you, and you only have to care about the version of the toolkit, rather than its components;
  - this includes versioning for different platforms btw, as a toolkit should also be usable with Scala.js or Scala Native, and people do get lost with where those `:` should go and in what numbers 😅 
- it's all contained in a single directive;
  - btw. there's also a [test scope toolkit directive ](https://scala-cli.virtuslab.org/docs/guides/introduction/toolkit#scala-toolkit-and-tests) in `//> using test.toolkit`;
- if some pieces of the toolkit should only land in the `test` scope, it also happens behind the curtain, where the `toolkit-test` dependencies will only be present on the test scope classpath.

Hope that helps!

**Reply by Gedochao** [reply]
BTW side note, the `latest` version for Toolkit directives is likely going to be deprecated in favour of `default` in the next Scala CLI release. For more information refer to https://github.com/VirtusLab/scala-cli/pull/2622
Probably irrelevant for a custom toolkit, but mentioning it just in case it affects for your talk.

**Reply by yadavan88** [reply]
@Gedochao Thank you so much for detailed information. I got more clarity now :)

**Reply by yadavan88** [reply]
@Gedochao One more question. I am trying to use the custom published toolkit to write test. In this case, it is using test. However, when I use it, it doesn't resolve the library:

```
//> using test.toolkit com.yadavan88:0.1.0
import utest._
object CustomToolkitTest extends TestSuite {
  val tests = Tests {
    test("simple test") {
      println("executing test... ")
    }
  }
}

```
Is the way I am adding the directive wrong?

**Reply by Gedochao** [reply]
@yadavan88 the directive looks correct (and if it wasn't, the error message should point you directly to the directive line)

Random guess... Is your test source file correctly identified as a test scope source? 
Refer [here](https://scala-cli.virtuslab.org/docs/commands/test#test-sources) for what I mean.
The test libraries within your toolkit should only be added to the test scope class path.
This may be confusing as the `test.*` directives can be declared anywhere in the project, including main scope sources.

**Reply by yadavan88** [reply]
My bad! Thank you so much @Gedochao 
I didn't use `.test.scala` notation. It works now
